File: app/src/main/java/com/aracroproducts/attention/AppServer.java
package com.aracroproducts.attention;

import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;


public class AppServer extends JobIntentService {
    private final String TAG = getClass().getName();

    protected static final String ACTION_POST_TOKEN = "com.aracroproducts.attention.action.token";
    protected static final String ACTION_SEND_ALERT = "com.aracroproducts.attention.action.send";

    protected static final String EXTRA_TOKEN = "com.aracroproducts.attention.extra.token";
    protected static final String EXTRA_ID = "com.aracroproducts.attention.extra.id";
    protected static final String EXTRA_TO = "com.aracroproducts.attention.extra.to";
    protected static final String EXTRA_FROM = "com.aracroproducts.attention.extra.from";
    protected static final String EXTRA_MESSAGE = "com.aracroproducts.attention.extra.message";

    protected static final String EXTRA_DATA = "com.aracroproducts.attention.extra.data";
    protected static final String EXTRA_RESULT_CODE = "com.aracroproducts.attention.extra.result_code";

    protected static final int CODE_SUCCESS = 0;
    protected static final int CODE_ERROR = 1;
    protected static final int CODE_NA = -1;

    private static final int JOB_ID = 0;

    private static final String PARAM_FUNCTION_ID = "post_id";
    private static final String PARAM_FUNCTION_ALERT = "send_alert";

    private static final String BASE_URL = "https://aracroproducts.com/attention/api/api.php?function=";

    public AppServer() {
        super();
    }

    public static void enqueueWork(Context context, Intent intent) {
        enqueueWork(context, AppServer.class, JOB_ID, intent);
    }

    @Override
    protected void onHandleWork(@NonNull Intent intent) {
        Log.d(TAG, "Handling work");
        if (intent.getAction() != null) {
            Log.d(TAG, intent.toString() + " Action: " + intent.getAction());
            final String action = intent.getAction();
            LocalBroadcastManager manager = LocalBroadcastManager.getInstance(this);

            //PendingIntent reply = intent.getParcelableExtra(EXTRA_PENDING_RESULT);
            Intent result = new Intent();
            switch (action) {
                case ACTION_POST_TOKEN:
                    String token = intent.getStringExtra(EXTRA_TOKEN);
                    String id = intent.getStringExtra(EXTRA_ID);
                    result.setAction(ACTION_POST_TOKEN);
                    boolean success = connect(true, new String[][]{{"token", token}, {"id", id}}, PARAM_FUNCTION_ID);
                    if (success) {
                        Log.d(TAG, "Message sent");
                        result.putExtra(EXTRA_DATA, "Sent id successfully");
                        result.putExtra(EXTRA_RESULT_CODE, CODE_SUCCESS);
                    } else {
                        Log.e(TAG, "An error occurred");
                        result.putExtra(EXTRA_DATA, "Error sending ID");
                        result.putExtra(EXTRA_RESULT_CODE, CODE_ERROR);
                    }
                    break;
                case ACTION_SEND_ALERT:
                    String to = intent.getStringExtra(EXTRA_TO);
                    String from = intent.getStringExtra(EXTRA_FROM);
                    String message = intent.getStringExtra(EXTRA_MESSAGE);
                    success = connect(true, new String[][]{{"to", to}, {"from", from}, {"message", message}}, PARAM_FUNCTION_ALERT);
                    result.setAction(ACTION_SEND_ALERT);

                    if (success) {
                        Log.d(TAG, "Alert sent");
                        result.putExtra(EXTRA_DATA, "Sent alert successfully");
                        result.putExtra(EXTRA_RESULT_CODE, CODE_SUCCESS);
                    } else {
                        Log.e(TAG, "An error occurred");
                        result.putExtra(EXTRA_DATA, "Error sending alert");
                        result.putExtra(EXTRA_RESULT_CODE, CODE_ERROR);
                    }
                    //todo provide real callbacks
                    //todo figure out a good way to pass result back (implement a BroadcastSender, probably)
            }

            manager.sendBroadcast(result);

        }
    }

    /**
     * Sends data via http
     * @param post      - Whether to use POST or GET
     * @param params    - Additional parameters for the server
     * @param function  - Which function it is calling (one of 'post_id' or 'send_alert')
     * @return          - A boolean representing whether the request was successful
     */
    private boolean connect(@SuppressWarnings("SameParameterValue") boolean post, String[][] params, String function) {
        HttpURLConnection c;
        String getParams = parseGetParams(params, !post);
        try {
            URL url;
            if (post) {
                url = new URL(BASE_URL + function);
            } else {
                url = new URL(BASE_URL + function + getParams);
            }
            c = (HttpURLConnection) url.openConnection();
            if (post) {
                c.setRequestMethod("POST");
            } else {
                c.setRequestMethod("GET");
            }
            c.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            c.setRequestProperty("charset", "utf-8");
            c.setRequestProperty("User-Agent", "Attention! client app for Android");
            c.setRequestProperty("Cache-control", "no-cache"); // tell server not to send cached response
            c.setUseCaches(false); // do it twice just to be sure

            if (post) {
                c.setDoOutput(true);
                DataOutputStream os = new DataOutputStream(c.getOutputStream()); // "out" is considered what the app is sending to the server
                os.write(getParams.getBytes(StandardCharsets.UTF_8));
                os.flush();
                os.close();
            }
            int code = c.getResponseCode();
            BufferedReader br = new BufferedReader(new InputStreamReader(c.getInputStream(), StandardCharsets.UTF_8)); // likewise, "in" is what is coming from the server
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) {
                sb.append(line);
            }
            br.close();
            Log.d(TAG, url + " " + code + " " + sb.toString());

            return code == HttpURLConnection.HTTP_CREATED || code == HttpURLConnection.HTTP_OK;

        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    private String parseGetParams(String[][] params, boolean leadingAmp) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < params.length; i++) {
            if (i != 0 || leadingAmp) {
                builder.append('&');
            }
            builder.append(params[i][0]);
            builder.append('=');
            builder.append(params[i][1]);
        }
        return builder.toString();
    }

}

File: app/src/main/java/com/aracroproducts/attention/AppServer.kt
package com.aracroproducts.attention

import android.content.Context
import androidx.core.app.JobIntentService
import android.content.Intent
import android.util.Log
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import java.io.BufferedReader
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStreamReader
import java.lang.StringBuilder
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

// TODO
open class AppServer : JobIntentService() {
    private val sTAG = javaClass.name
    override fun onHandleWork(intent: Intent) {
        Log.d(sTAG, "Handling work")
        if (intent.action != null) {
            Log.d(sTAG, intent.toString() + " Action: " + intent.action)
            val action = intent.action
            val manager = LocalBroadcastManager.getInstance(this)

            //PendingIntent reply = intent.getParcelableExtra(EXTRA_PENDING_RESULT);
            val result = Intent()
            when (action) {
                ACTION_POST_TOKEN -> {
                    val token = intent.getStringExtra(EXTRA_TOKEN)
                    val id = intent.getStringExtra(EXTRA_ID)
                    result.action = ACTION_POST_TOKEN
                    val success = connect(true, arrayOf(arrayOf("token", token), arrayOf("id", id)), PARAM_FUNCTION_ID)
                    if (success) {
                        Log.d(sTAG, "Message sent")
                        result.putExtra(EXTRA_DATA, "Sent id successfully")
                        result.putExtra(EXTRA_RESULT_CODE, CODE_SUCCESS)
                    } else {
                        Log.e(sTAG, "An error occurred")
                        result.putExtra(EXTRA_DATA, "Error sending ID")
                        result.putExtra(EXTRA_RESULT_CODE, CODE_ERROR)
                    }
                }
                ACTION_SEND_ALERT -> {
                    val to = intent.getStringExtra(EXTRA_TO)
                    val from = intent.getStringExtra(EXTRA_FROM)
                    val message = intent.getStringExtra(EXTRA_MESSAGE)
                    val success = connect(true, arrayOf(arrayOf("to", to), arrayOf("from", from), arrayOf("message", message)), PARAM_FUNCTION_ALERT)
                    result.action = ACTION_SEND_ALERT
                    if (success) {
                        Log.d(sTAG, "Alert sent")
                        result.putExtra(EXTRA_DATA, "Sent alert successfully")
                        result.putExtra(EXTRA_RESULT_CODE, CODE_SUCCESS)
                    } else {
                        Log.e(sTAG, "An error occurred")
                        result.putExtra(EXTRA_DATA, "Error sending alert")
                        result.putExtra(EXTRA_RESULT_CODE, CODE_ERROR)
                    }
                }
            }
            manager.sendBroadcast(result)
        }
    }

    /**
     * Sends data via http
     * @param post      - Whether to use POST or GET
     * @param params    - Additional parameters for the server
     * @param function  - Which function it is calling (one of 'post_id' or 'send_alert')
     * @return          - A boolean representing whether the request was successful
     */
    private fun connect(post: Boolean, params: Array<Array<String?>>, function: String): Boolean {
        val c: HttpURLConnection
        val getParams = parseGetParams(params, !post)
        try {
            val url: URL = if (post) {
                URL(BASE_URL + function)
            } else {
                URL(BASE_URL + function + getParams)
            }
            c = url.openConnection() as HttpURLConnection
            if (post) {
                c.requestMethod = "POST"
            } else {
                c.requestMethod = "GET"
            }
            c.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
            c.setRequestProperty("charset", "utf-8")
            c.setRequestProperty("User-Agent", "Attention! client app for Android")
            c.setRequestProperty("Cache-control", "no-cache") // tell server not to send cached response
            c.useCaches = false // do it twice just to be sure
            if (post) {
                c.doOutput = true
                val os = DataOutputStream(c.outputStream) // "out" is considered what the app is sending to the server
                os.write(getParams.toByteArray(StandardCharsets.UTF_8))
                os.flush()
                os.close()
            }
            val code = c.responseCode
            val br = BufferedReader(InputStreamReader(c.inputStream, StandardCharsets.UTF_8)) // likewise, "in" is what is coming from the server
            val sb = StringBuilder()
            var line: String?
            while (br.readLine().also { line = it } != null) {
                sb.append(line)
            }
            br.close()
            Log.d(sTAG, "$url $code $sb")
            return code == HttpURLConnection.HTTP_CREATED || code == HttpURLConnection.HTTP_OK
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return false
    }

    private fun parseGetParams(params: Array<Array<String?>>, leadingAmp: Boolean): String {
        val builder = StringBuilder()
        for (i in params.indices) {
            if (i != 0 || leadingAmp) {
                builder.append('&')
            }
            builder.append(params[i][0])
            builder.append('=')
            builder.append(params[i][1])
        }
        return builder.toString()
    }

    companion object {
        const val ACTION_POST_TOKEN = "com.aracroproducts.attention.action.token"
        const val ACTION_SEND_ALERT = "com.aracroproducts.attention.action.send"
        const val EXTRA_TOKEN = "com.aracroproducts.attention.extra.token"
        const val EXTRA_ID = "com.aracroproducts.attention.extra.id"
        const val EXTRA_TO = "com.aracroproducts.attention.extra.to"
        const val EXTRA_FROM = "com.aracroproducts.attention.extra.from"
        const val EXTRA_MESSAGE = "com.aracroproducts.attention.extra.message"
        const val EXTRA_DATA = "com.aracroproducts.attention.extra.data"
        const val EXTRA_RESULT_CODE = "com.aracroproducts.attention.extra.result_code"
        const val CODE_SUCCESS = 0
        const val CODE_ERROR = 1
        const val CODE_NA = -1
        private const val JOB_ID = 0
        private const val PARAM_FUNCTION_ID = "post_id"
        private const val PARAM_FUNCTION_ALERT = "send_alert"
        private const val BASE_URL = "https://aracroproducts.com/attention/api/api.php?function="
        fun enqueueWork(context: Context?, intent: Intent?) {
            enqueueWork(context!!, AppServer::class.java, JOB_ID, intent!!)
        }
    }
}